### ISSUE TITLE
TradeCortex

### ISSUE THREAD

**Issue by Urbano128** [Urbano128/CortexIA#3]
import math
import random
import json
import numpy as np
import joblib
import matplotlib.pyplot as plt
from flask import Flask, request, jsonify
from sklearn.linear_model import LinearRegression
from textblob import TextBlob
from sqlalchemy import create_engine, Column, Integer, Float, String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

# -------------------------------
# Configuración de la Base de Datos para ejemplos de entrenamiento
# -------------------------------
DATABASE_URL = "sqlite:///training.db"
Base = declarative_base()

class TrainingExample(Base):
    __tablename__ = "training_examples"
    id = Column(Integer, primary_key=True, autoincrement=True)
    features = Column(String, nullable=False)  # Se almacena en formato JSON
    target = Column(Float, nullable=False)

engine = create_engine(DATABASE_URL, echo=False)
Base.metadata.create_all(engine)
SessionLocal = sessionmaker(bind=engine)

# -------------------------------
# MODELO DE APRENDIZAJE AUTOMÁTICO (ML)
# -------------------------------
class MLModel:
    def __init__(self):
        self.model = LinearRegression()
        self.trained = False
        self.model_file = "modelo_cortexIA.pkl"

    def entrenar_modelo(self, datos_X, datos_y):
        X = np.array(datos_X)
        y = np.array(datos_y)
        self.model.fit(X, y)
        self.trained = True
        joblib.dump(self.model, self.model_file)

    def predecir(self, features):
        if not self.trained:
            try:
                self.model = joblib.load(self.model_file)
                self.trained = True
            except Exception as e:
                return features[0]  # Fallback si hay error
        return self.model.predict(np.array([features]))[0]

# -------------------------------
# CLASE Cryptolight (Análisis estructural)
# -------------------------------
class Cryptolight:
    def analizar_estructura(self, valores):
        valores_np = np.array(valores)
        simetria = np.sum(valores_np % 2 == 0) / len(valores)
        disimetria = np.sum(valores_np % 2 != 0) / len(valores)
        recomendacion = self._recomendar_reto(simetria, disimetria)
        return {
            "valores": valores,
            "simetria": round(simetria, 2),
            "disimetria": round(disimetria, 2),
            "recomendacion": recomendacion
        }

    def _recomendar_reto(self, sim, dis):
        if sim > 0.7:
            return "Enfócate en retos estructurados para estabilidad."
        elif dis > 0.7:
            return "Explora soluciones no convencionales."
        return "Alterna entre ambos enfoques."

# -------------------------------
# CLASE CortexIA (Integración del motor, ML, autoaprendizaje, patrón complejo y chat emocional)
# -------------------------------
class CortexIA:
    def __init__(self):
        self.cryptolight = Cryptolight()
        self.ml_model = MLModel()

    def motor_simetrico_inverso(self, market_inputs):
        valores_np = np.array(market_inputs)
        mean_prediction = np.mean(valores_np)
        angles = valores_np % 360
        mean_angle = np.mean(angles)
        sigma = np.std(angles)
        # Filtrado de outliers
        filtered_angles = angles[np.abs(angles - mean_angle) < 30]
        filtered_mean = np.mean(filtered_angles) if len(filtered_angles) > 0 else mean_angle
        patron_simetrico_alterno = np.tanh(sigma) * math.pi + filtered_mean
        quantum_assimilation = 4 * random.random() * (1 - random.random())
        final_prediction = np.mean([mean_prediction, patron_simetrico_alterno, quantum_assimilation])
        return final_prediction

    def patron_ordenanza_tecnica_mezclable(self, market_inputs):
        base_prediction = self.motor_simetrico_inverso(market_inputs)
        analisis = self.cryptolight.analizar_estructura(market_inputs)
        features = [base_prediction, analisis["simetria"], analisis["disimetria"]]
        ml_correction = self.ml_model.predecir(features)
        blended_prediction = 0.6 * base_prediction + 0.4 * ml_correction
        return blended_prediction, analisis

    def patron_pregunta_compleja_math(self, market_inputs):
        # Patrón simétrico (números pares)
        evens = [x for x in market_inputs if x % 2 == 0]
        simetrico_calc = np.mean(evens) if evens else 0
        # Patrón dismétrico (números impares)
        odds = [x for x in market_inputs if x % 2 != 0]
        dismetrico_calc = np.mean(odds) if odds else 0
        # Patrón simétrico total (promedio)
        simetrico_total = np.mean(market_inputs) if market_inputs else 0

        # Generación de gráfica para la estructura visual
        plt.figure(figsize=(8, 5))
        plt.plot(market_inputs, 'o-', label="Market Inputs")
        plt.axhline(y=simetrico_calc, color='red', linestyle='--', label="Simétrico (Pares)")
        plt.axhline(y=dismetrico_calc, color='green', linestyle='--', label="Dismétrico (Impares)")
        plt.axhline(y=simetrico_total, color='blue', linestyle='--', label="Simétrico Total")
        plt.title("Estructura Visual del Patrón Complejo")
        plt.xlabel("Índice")
        plt.ylabel("Valor")
        plt.legend()
        estructura_filename = "estructura_visual.png"
        plt.savefig(estructura_filename)
        plt.close()

        promedio_combinado = (simetrico_calc + dismetrico_calc + simetrico_total) / 3
        resultado_motor20 = promedio_combinado * 20

        return {
            "simetrico_calc": simetrico_calc,
            "dismetrico_calc": dismetrico_calc,
            "simetrico_total": simetrico_total,
            "estructura_visual": estructura_filename,
            "resultado_motor20": resultado_motor20
        }

    def actualizar_autoaprendizaje(self, features, target):
        """
        Guarda el ejemplo de entrenamiento en la base de datos.
        Cuando se acumulan 5 o más ejemplos, se entrena el modelo ML y se limpia la base.
        """
        db = SessionLocal()
        example = TrainingExample(features=json.dumps(features), target=target)
        db.add(example)
        db.commit()
        count = db.query(TrainingExample).count()
        if count >= 5:
            examples = db.query(TrainingExample).all()
            train_features = [json.loads(ex.features) for ex in examples]
            train_targets = [ex.target for ex in examples]
            self.ml_model.entrenar_modelo(train_features, train_targets)
            db.query(TrainingExample).delete()
            db.commit()
        db.close()

    def evaluar_idea(self, idea_texto, estructura_numerica):
        blended_prediction, analisis = self.patron_ordenanza_tecnica_mezclable(estructura_numerica)
        decision = "Aceptada" if analisis["simetria"] >= 0.5 else "Requiere ajuste"
        return {
            "idea": idea_texto,
            "estructura": analisis,
            "prediccion_cortexIA": round(blended_prediction, 4),
            "decision": decision
        }

    def chat_emocional_simetrico(self, message):
        sentimiento = TextBlob(message).sentiment.polarity
        if sentimiento > 0.3:
            response = (
                "Me alegra que estés bien. La simetría en tus emociones genera armonía. ¿Te gustaría contarme "
                "más aspectos positivos de tu día?"
            )
        elif sentimiento < -0.3:
            response = (
                "Lamento que estés pasando por momentos difíciles. La simetría en la adversidad puede revelar equilibrio "
                "y oportunidades de mejora. ¿Quieres compartir lo que te preocupa?"
            )
        else:
            response = (
                "Percibo un estado neutral, lo que denota estabilidad. A veces, ese balance invita a la reflexión. ¿Te animas a explorar "
                "algún aspecto reciente que te haya marcado?"
            )
        return response

# -------------------------------
# SERVIDOR FLASK CON TODOS LOS ENDPOINTS
# -------------------------------
app = Flask(__name__)
cortex = CortexIA()

@app.route("/interact", methods=["POST"])
def interact_with_cortexia():
    data = request.get_json()
    idea_query = data.get("idea", "").strip()
    if not idea_query:
        return jsonify({"error": "No se ingresó ninguna idea."}), 400
    # Estructura numérica de ejemplo (simulación de datos de mercado)
    estructura_defecto = [100, 105, 95, 102]
    resultado = cortex.evaluar_idea(idea_query, estructura_defecto)
    return jsonify(resultado)

@app.route("/feedback", methods=["POST"])
def feedback():
    data = request.get_json()
    features = data.get("features")
    target = data.get("target")
    if features is None or target is None:
        return jsonify({"error": "Se requieren 'features' y 'target'."}), 400
    cortex.actualizar_autoaprendizaje(features, target)
    return jsonify({"status": "Modelo actualizado con éxito."})

@app.route("/complex_pattern", methods=["POST"])
def complex_pattern():
    data = request.get_json()
    market_inputs = data.get("market_inputs")
    if not market_inputs or not isinstance(market_inputs, list):
        return jsonify({"error": "Se requiere 'market_inputs' como lista de números."}), 400
    result = cortex.patron_pregunta_compleja_math(market_inputs)
    return jsonify(result)

@app.route("/chat_humano", methods=["POST"])
def chat_humano():
    data = request.get_json()
    message = data.get("message", "").strip()
    if not message:
        return jsonify({"error": "No se ingresó ningún mensaje."}), 400
    response = cortex.chat_emocional_simetrico(message)
    return jsonify({"respuesta": response})

# -------------------------------
# Nota sobre escalabilidad:
# En producción, se recomienda desplegar la aplicación con Gunicorn o Uvicorn para manejar múltiples peticiones.
# Ejemplo: gunicorn -w 4 app:app
# -------------------------------

if __name__ == "__main__":
    app.run(debug=True)
